### PR TITLE
Add patchelf/0.12

### PR DESCRIPTION
**Pull request by cbachhuber**
Specify library name and version:  **patchelf/0.12**

[PatchELF](https://github.com/NixOS/patchelf) is used in our private conan project. We deem it useful for the general public, so everyone profits from having this in CCI.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/patchelf/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    def _configure_autotools(self):
+        if self._autotools:
+            return self._autotools
+        self._autotools = AutoToolsBuildEnvironment(self)
+        self._autotools.configure(configure_dir=self._source_subfolder)
+        return self._autotools
+
+    def build(self):
+        with tools.chdir(self._source_subfolder):
+            self.run("{} -fiv --warnings=all".format(tools.get_env("AUTORECONF")), run_environment=True)
```
Comment (cbachhuber) [inline recipes/patchelf/all/conanfile.py:41]: I've added `--warnings=all` from patchelf's [build script](https://github.com/NixOS/patchelf/blob/master/bootstrap.sh) to be consistent with patchelf's build process. This flag is not strictly required, though. We can discuss removing it.

**Review by blackliner — DISMISSED** [review]

**Review by SSE4 — APPROVED** [review]

File: recipes/patchelf/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    def build_requirements(self):
+        self.build_requires("libtool/2.4.6")
+
+    def configure(self):
+        if not tools.is_apple_os(self.settings.os) and self.settings.os != "Linux":
```
Comment (madebr) [inline recipes/patchelf/all/conanfile.py:26]: `!apple && !linux` is equivalent with `!linux`.
```suggestion
        if self.settings.os not in ("FreeBSD", "Linux"):
```
I added `FreeBSD` to the suggestion.
Don't know whether it works there, but perhaps?

File: recipes/patchelf/all/conanfile.py (hunk 1 above)
Comment (madebr) [reply recipes/patchelf/all/conanfile.py:41]: As long as there's no `--warnings=error` :smile:

**Review by madebr — COMMENTED** [review]
> [PatchELF](https://github.com/NixOS/patchelf) is used in our private conan project. We deem it useful for the general public, so everyone profits from having this in CCI.

Thank you very much!

**Review by intelligide — APPROVED** [review]

File: recipes/patchelf/all/conanfile.py (hunk 2 above)
Comment (cbachhuber) [reply recipes/patchelf/all/conanfile.py:26]: [Looks good](https://www.freshports.org/sysutils/patchelf), thanks for the suggestion! 👍

File: recipes/patchelf/all/test_package/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+from conans import ConanFile, CMake, tools
+
+
+class TestPackageConan(ConanFile):
+    settings = "os", "compiler", "build_type", "arch"
+
+    def test(self):
+        if not tools.cross_building(self.settings, skip_x64_x86=True):
```
Comment (madebr) [inline recipes/patchelf/all/test_package/conanfile.py:8]:
```suggestion
        if not tools.cross_building(self, skip_x64_x86=True):
```

Comment (prince-chrismc) [reply recipes/patchelf/all/test_package/conanfile.py:8]: https://github.com/conan-io/hooks/issues/324 spreading the word!

File: recipes/patchelf/all/conanfile.py (hunk 1 above)
Comment (prince-chrismc) [reply recipes/patchelf/all/conanfile.py:41]: @madebr Random question but I noticed the usage of autoreconf to be really varied...
https://github.com/conan-io/conan-center-index/search?p=3&q=AUTORECONF

Things like https://github.com/conan-io/conan-center-index/blob/41d4fbb8c218f9d938ccb8b9d888bff52002e62e/recipes/cunit/all/conanfile.py#L112 just look wrong and https://github.com/conan-io/conan-center-index/blob/d28a79b253a674e86e7296c7daef89336bd4f96a/recipes/libsodium/all/conanfile.py#L95
doesn't match the reset

❓ 
- `tools.get_env` vs `os.environ`
- is `run_environment=True` required/helpful?
- is `win_bash=tools.os_info.is_windows` required for building on Windows?

➡️ I want to update our "how to add packages" docs with the copy gnuconfig stuff and this but I am not sure what a good example looks like

Comment (madebr) [reply recipes/patchelf/all/conanfile.py:41]: Here is the man page of `autoreconf`: https://linux.die.net/man/1/autoreconf
`-f` is to "force" rewrite all files (mark all files as out of date) and `-v` is for verbose output.
So `autoreconf -i` should be enough. I've started using `-fiv` just to be sure. I've encountered these options also in random utility scripts of projects to re-create their configure scripts.

- We've started to converge around `tools.get_env("AUTORECONF")`. Other usages are from the "old times" where that utility function did not exist yet or from copy-paste.
A contra for using `tools.get_env("AUTORECONF")` is that when `autoconf` is not a build requirement when cross building, then it tries to run `None -fiv` instead.
- `run_environment=True` might be useful: autoconf is a combination of perl and m4 scripts. Both can run executables.
Or perl might be a conan package, needing a run environment. (I've started working on a perl recipe, but it's unfinished)
- `win_bash=tools.os_info.is_windows` is mostly needed to interpret the [shebang at the top of autoreconf](https://github.com/autotools-mirror/autoconf/blob/64df9b4523fe3b3c5d6e444fb65b1d92fc5073ce/bin/autoreconf.in#L1). Autoconf is a GNU project that might expect forward slashes and other unix tools present. I don't think a bash script is run. I'm not sure though. It runs fine on Windows with strawberryperl.

**Review by blackliner — APPROVED** [review]

File: recipes/patchelf/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    _autotools = None
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    def build_requirements(self):
+        self.build_requires("libtool/2.4.6")
+
+    def configure(self):
```
Comment (uilianries) [inline recipes/patchelf/all/conanfile.py:25]:
```suggestion
    def validate(self):
```

**Review by madebr — APPROVED** [review]

**Review by intelligide — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Pull request merged**